src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs

```diff
@@ -405,8 +421,259 @@ static void M1()
+        public void QueryOnDefault()
+        {
+            string source =
+@"static class C
+{
+    static void Main()
+    {
+        var q = from x in default select x;
+    }
+}
+";
+            var compilation = CreateCompilationWithMscorlibAndSystemCore(source, parseOptions: TestOptions.ExperimentalParseOptions);
+            compilation.VerifyDiagnostics(
+                // (5,35): error CS0186: Use of null is not valid in this context
+                //         var q = from x in default select x;
+                Diagnostic(ErrorCode.ERR_NullNotValid, "select x").WithLocation(5, 35)
+                );
+        }
+
+        [Fact]
+        public void DefaultInConditionalExpression()
+        {
+            string source =
+@"static class C
+{
+    static void Main()
+    {
+        var x = default ? 4 : 5;
+        System.Console.Write(x);
+    }
+}
+";
+            var compilation = CreateCompilationWithMscorlibAndSystemCore(source, parseOptions: TestOptions.ExperimentalParseOptions, options: TestOptions.DebugExe);
+            compilation.VerifyDiagnostics();
+            CompileAndVerify(compilation, expectedOutput: "5");
+        }
+
+        [Fact]
+        public void AlwaysNonNull()
+        {
+            string source =
+@"static class C
+{
+    static void Main()
+    {
+        System.Console.Write((int?)1 == default);
+        System.Console.Write(default == (int?)1);
+    }
+}
+";
+            var compilation = CreateCompilationWithMscorlibAndSystemCore(source, parseOptions: TestOptions.ExperimentalParseOptions);
+            compilation.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void ThrowDefault()
+        {
+            var text = @"
+class C
+{
+    static void Main()
+    {
+        throw default;
+    }
+}";
+
+            var comp = CreateCompilationWithMscorlib(text, parseOptions: TestOptions.ExperimentalParseOptions, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics(
+                // (6,15): error CS0155: The type caught or thrown must be derived from System.Exception
+                //         throw default;
+                Diagnostic(ErrorCode.ERR_BadExceptionType, "default").WithLocation(6, 15)
+                );
+        }
+
+        [Fact]
+        public void DefaultInAsOperator()
+        {
+            var text = @"
+class C
+{
+    static void Main()
+    {
+        System.Console.Write(default as long);
+    }
+}";
+
+            var comp = CreateCompilationWithMscorlib(text, parseOptions: TestOptions.ExperimentalParseOptions, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics(
+                // (6,30): error CS0077: The as operator must be used with a reference type or nullable type ('long' is a non-nullable value type)
+                //         System.Console.Write(default as long);
+                Diagnostic(ErrorCode.ERR_AsMustHaveReferenceType, "default as long").WithArguments("long").WithLocation(6, 30)
+                );
```

Comment:
`InvocationOnDefault` has `nameof(default)` (illegal). `DefaultInAsOperator` has `default as long` (illegal).
I'll add tests for the other suggestions.